include/cuco/detail/storage.cuh

```diff
@@ -0,0 +1,413 @@
+   *
+   * @param windows Pointer to the windows array
+   * @param num_windows Number of slots
+   */
+  explicit constexpr aow_storage_ref(Extent num_windows, window_type* windows) noexcept
+    : aow_storage_base<WindowSize, T, Extent>{num_windows}, windows_{windows}
+  {
+  }
+
+  /**
+   * @brief Gets windows array.
+   *
+   * @return Pointer to the first window
+   */
+  [[nodiscard]] __device__ inline constexpr window_type* windows() noexcept { return windows_; }
+
+  /**
+   * @brief Gets windows array.
+   *
+   * @return Pointer to the first window
+   */
+  [[nodiscard]] __device__ inline constexpr window_type* windows() const noexcept
+  {
+    return windows_;
+  }
+
+  /**
+   * @brief Returns an array of elements (window) for a given index.
+   *
+   * @param index Index of the first element of the window
+   * @return An array of elements
+   */
+  [[nodiscard]] __device__ inline constexpr window_type window(size_type index) const noexcept
+  {
+    return *reinterpret_cast<window_type*>(
+      __builtin_assume_aligned(windows_ + index, sizeof(value_type) * window_size));
+  }
+
+ private:
+  window_type* windows_;  ///< Pointer to the windows array
+};
+
+/**
+ * @brief Array of window structure open addressing storage class.
+ *
+ * @tparam WindowSize Number of slots in each window
+ * @tparam T struct type
+ * @tparam Extent Type of extent denoting number of windows
+ * @tparam Allocator Type of allocator used for device storage
+ */
+template <int32_t WindowSize, typename T, typename Extent, typename Allocator>
+class aow_storage : public aow_storage_base<WindowSize, T, Extent> {
+ public:
+  using base_type = aow_storage_base<WindowSize, T, Extent>;  ///< AoW base class type
+
+  using base_type::window_size;  ///< Number of elements processed per window
+
+  using extent_type = typename base_type::extent_type;  ///< Storage extent type
+  using size_type   = typename base_type::size_type;    ///< Storage size type
+  using value_type  = typename base_type::value_type;   ///< Storage element type
+  using window_type = typename base_type::window_type;  ///< Window storage type
+
+  using base_type::capacity;
+  using base_type::num_windows;
+
+  using allocator_type =
+    typename std::allocator_traits<Allocator>::rebind_alloc<window_type>;  ///< Type of the
+                                                                           ///< allocator to
+                                                                           ///< (de)allocate windows
+  using window_deleter_type = custom_deleter<allocator_type>;  ///< Type of window deleter
+  using ref_type = aow_storage_ref<window_size, value_type, extent_type>;  ///< Storage ref type
+
+  /**
+   * @brief Constructor of AoW storage.
+   *
+   * @param size Number of slots to (de)allocate
+   * @param allocator Allocator used for (de)allocating device storage
+   */
+  explicit constexpr aow_storage(Extent size, Allocator const& allocator)
```

Comment:
Storage shouldn't be aware of the concept of probing scheme (or CG size). Closing this since `make_valid_extent` should be exclusively used to determine the capacity of the map/set or storage.